### PR TITLE
chore: remove all raw packet reading from chat packet handler 

### PR DESCRIPTION
**Pull request by aronwk-aaron**
and general chat packet cleanup

Tested that things still work

### CODE REVIEW

**Review by Xiphoseer — DISMISSED** [review]

**Review by Xiphoseer — APPROVED** [review]

File: dChatServer/ChatPacketHandler.cpp [hunk 1]
```diff
@@ -132,8 +125,8 @@ void ChatPacketHandler::HandleFriendRequest(Packet* packet) {
 	// If at this point we dont have a target, then they arent online and we cant send the request.
 	// Send the response code that corresponds to what the error is.
 	if (!requestee) {
-		requestee.playerName = playerName;
-		auto responseType = Database::Get()->GetCharacterInfo(playerName)
+		requestee.playerName = playerName.GetAsString();
```
Comment (EmosewaMC) [inline dChatServer/ChatPacketHandler.cpp:128]: There is a lot of copying of the LUWString.GetAsString going on here.  Consider a `const auto nameAsStr = playerName.GetAsString();`

File: dChatServer/ChatPacketHandler.cpp [hunk 2]
```diff
@@ -288,13 +285,15 @@ void ChatPacketHandler::HandleFriendResponse(Packet* packet) {
...
 	inStream.Read(playerID);
-	std::string friendName = PacketUtils::ReadString(0x14, packet, true);
+	inStream.IgnoreBytes(4);
+	inStream.Read(friendName);
 
 	//we'll have to query the db here to find the user, since you can delete them while they're offline.
 	//First, we need to find their ID:
 	LWOOBJID friendID = 0;
-	auto friendIdResult = Database::Get()->GetCharacterInfo(friendName);
+	auto friendIdResult = Database::Get()->GetCharacterInfo(friendName.GetAsString());
```
Comment (EmosewaMC) [inline dChatServer/ChatPacketHandler.cpp:296]: Same copy twice happens here, again consider a const declaration and just copy it once.

File: dChatServer/ChatPacketHandler.cpp [hunk 3]
```diff
@@ -337,121 +337,142 @@ void ChatPacketHandler::HandleRemoveFriend(Packet* packet) {
...
-	inStream.SetReadOffset(0x14 * 8);
-
-	uint8_t channel = 0;
+	eChatChannel channel;
+	uint32_t size;
+	
+	inStream.IgnoreBytes(4);
 	inStream.Read(channel);
+	inStream.Read(size);
+	inStream.IgnoreBytes(77);
```
Comment (EmosewaMC) [inline dChatServer/ChatPacketHandler.cpp:352]: Why 77?  This should be commented and elaborated on for future maintainers otherwise we'll RE this again.

File: dChatServer/ChatPacketHandler.cpp [hunk 4]
```diff
@@ -337,121 +337,142 @@ void ChatPacketHandler::HandleRemoveFriend(Packet* packet) {
...
-	std::string message = PacketUtils::ReadString(0x66, packet, true, 512);
-
-	LOG("Got a message from (%s) [%d]: %s", sender.playerName.c_str(), channel, message.c_str());
+	LUWString message(size);
+	inStream.Read(message);
+	
+	LOG("Got a message from (%s) via [%s]: %s", sender.playerName.c_str(), StringifiedEnum::ToString(channel).data(), message.GetAsString().c_str());
 
-	if (channel != 8) return;
+	if (channel == eChatChannel::TEAM) {
```
Comment (EmosewaMC) [inline dChatServer/ChatPacketHandler.cpp:359]: Why was this inverted?  Bouncer was good for early return before to reduce scope.  If its to allow for more channels, consider a switch case instead for future use intent

File: dChatServer/ChatPacketHandler.cpp [hunk 5]
```diff
@@ -337,121 +337,142 @@ void ChatPacketHandler::HandleRemoveFriend(Packet* packet) {
...
-		SystemAddress sysAddr = otherMember.sysAddr;
-		SEND_PACKET;
-	}
-}
+	inStream.IgnoreBytes(4);
+	inStream.Read(channel);
+	inStream.Read(size);
+	inStream.IgnoreBytes(77);
+	inStream.Read(receiverName);
+	inStream.IgnoreBytes(2);
```
Comment (EmosewaMC) [inline dChatServer/ChatPacketHandler.cpp:408]: Same thing here with the magic numbers.  Try to elaborate on them if possible.

File: dChatServer/ChatPacketHandler.cpp [hunk 6]
```diff
@@ -337,121 +337,142 @@ void ChatPacketHandler::HandleRemoveFriend(Packet* packet) {
...
 
-	//Get the bois:
-	const auto& goonA = Game::playerContainer.GetPlayerData(senderID);
-	const auto& goonB = Game::playerContainer.GetPlayerData(receiverName);
-	if (!goonA || !goonB) return;
+	const auto& receiver = Game::playerContainer.GetPlayerData(receiverName.GetAsString());
+	if (!receiver) return; // TODO: respond with not online or error
 
-	if (goonA.GetIsMuted()) return;
+	// TODO: check if they are friends, only friends should be able to whisper/private chat
```
Comment (EmosewaMC) [inline dChatServer/ChatPacketHandler.cpp:418]: container has this data already.  Why was the muted check removed?

File: dChatServer/ChatServer.cpp [hunk 7]
```diff
@@ -295,7 +297,7 @@ void HandlePacket(Packet* packet) {
 			break;
 
 		default:
-			LOG("Unknown CHAT id: %i", int(packet->data[3]));
+			LOG("Unknown/Unhandled CHAT Message id: %s (%i)", StringifiedEnum::ToString(chat_message_type).data(), chat_message_type);
```
Comment (EmosewaMC) [inline dChatServer/ChatServer.cpp:300]: nit: add the explicitly unhandled ones as a fall through case statement such that if we plan to handle them, it logs Unhandled.

File: dChatServer/PlayerContainer.h [hunk 8]
```diff
@@ -42,6 +42,8 @@ struct PlayerData {
 	std::string playerName;
 	std::vector<FriendData> friends;
 	std::vector<IgnoreData> ignoredPlayers;
+	uint8_t GMLevel = 0;
```
Comment (EmosewaMC) [inline dChatServer/PlayerContainer.h:45]: Why is eGameMasterLevel not used here?  In any case this is unused in this PR and should be documenated as a TODO to implement

File: dChatServer/ChatPacketHandler.cpp (hunk 5 above)
Comment (EmosewaMC) [reply dChatServer/ChatPacketHandler.cpp:408]: > Same thing here with the magic numbers. Try to elaborate on them if possible.

If this code is also repeated, consider a function to deserialize and a struct to propagate the data between functions

File: dChatServer/ChatPacketHandler.cpp (hunk 1 above)
Comment (aronwk-aaron) [reply dChatServer/ChatPacketHandler.cpp:128]: yeah, true

File: dChatServer/ChatPacketHandler.cpp (hunk 3 above)
Comment (aronwk-aaron) [reply dChatServer/ChatPacketHandler.cpp:352]: there's no data there, so I'll make a comment to indicate as such

File: dChatServer/ChatPacketHandler.cpp (hunk 4 above)
Comment (aronwk-aaron) [reply dChatServer/ChatPacketHandler.cpp:359]: yeah, there can be more channels supported, so a switch case would work well here

File: dChatServer/ChatPacketHandler.cpp (hunk 5 above)
Comment (aronwk-aaron) [reply dChatServer/ChatPacketHandler.cpp:408]: yeah, was thinking about breaking it out and how to do that

File: dChatServer/ChatPacketHandler.cpp (hunk 6 above)
Comment (aronwk-aaron) [reply dChatServer/ChatPacketHandler.cpp:418]: mute check is done above, was a redundant check (will double check tho)

File: dChatServer/PlayerContainer.h (hunk 8 above)
Comment (aronwk-aaron) [reply dChatServer/PlayerContainer.h:45]: noted, was like 3am and i forgot we had an emun for it lol

File: dChatServer/ChatPacketHandler.cpp (hunk 6 above)
Comment (aronwk-aaron) [reply dChatServer/ChatPacketHandler.cpp:418]: Also just didn't do it yet cause I wanted to keep this PR from getting to big, but can do

File: dChatServer/ChatPacketHandler.cpp (hunk 5 above)
Comment (aronwk-aaron) [reply dChatServer/ChatPacketHandler.cpp:408]: erh, not the reading, but the writing.
the reading is just slightly different enough to be really annoying to consolidate to a common place like that

File: dChatServer/ChatPacketHandler.cpp (hunk 6 above)
Comment (aronwk-aaron) [reply dChatServer/ChatPacketHandler.cpp:418]: yeah, goona is sender and the sender is check if muted above, so redundant check

File: dChatServer/ChatPacketHandler.cpp (hunk 5 above)
Comment (aronwk-aaron) [reply dChatServer/ChatPacketHandler.cpp:408]: will consolidate the writing in a future PR

File: dChatServer/ChatPacketHandler.cpp (hunk 6 above)
Comment (EmosewaMC) [reply dChatServer/ChatPacketHandler.cpp:418]: the friends check isnt even much more code
```
bool isFriends = std::find(friends.begin(), friends.end(), friendId) != friends.end();
if (friends) {}
```

Comment (aronwk-aaron) [reply dChatServer/ChatPacketHandler.cpp:418]: yeah, I'm just doing it right this moment lolol

File: dChatServer/ChatPacketHandler.cpp [hunk 9]
```diff
@@ -346,45 +348,31 @@ void ChatPacketHandler::HandleChatMessage(Packet* packet) {
 	eChatChannel channel;
 	uint32_t size;
 	
-	inStream.IgnoreBytes(4);
+	inStream.IgnoreBytes(4); // padding/garbage
 	inStream.Read(channel);
 	inStream.Read(size);
-	inStream.IgnoreBytes(77);
+	inStream.IgnoreBytes(77); // padding/garbage
```
Comment (EmosewaMC) [inline dChatServer/ChatPacketHandler.cpp:354]: I would consider "the structure the client uses to send this packet is shared in many chat messages that are sent to the server.  Because of this, there is a large gap of unused data here" or something along those lines just because this is an abnormally large chunk of data to just call "padding"

File: dChatServer/ChatPacketHandler.cpp [hunk 10]
```diff
@@ -398,72 +386,83 @@ void ChatPacketHandler::HandlePrivateChatMessage(Packet* packet) {
...
 
-	//To the sender:
-	{
-		CBITSTREAM;
-		BitStreamUtils::WriteHeader(bitStream, eConnectionType::CHAT_INTERNAL, eChatInternalMessageType::ROUTE_TO_PLAYER);
-		bitStream.Write(sender.playerID);
+	if (sender.isFTP) {
+		SendPrivateChatMessage(sender, receiver, sender, message, eChatMessageResponseCode::SENDERFREETRIAL);
+		return;
+	}
```
Comment (EmosewaMC) [inline dChatServer/ChatPacketHandler.cpp:422]: I believe any logic actually involving FTP should be removed.  We shouldn't even have work setup to support it in dlu.

File: dChatServer/ChatPacketHandler.cpp [hunk 11]
```diff
@@ -398,72 +386,83 @@ void ChatPacketHandler::HandlePrivateChatMessage(Packet* packet) {
...
-		bitStream.Write(sender.playerID);
-		bitStream.Write<uint16_t>(0); // sourceID
-		bitStream.Write(sender.GMLevel);
-		bitStream.Write(LUWString(receiver.playerName));
-		bitStream.Write(receiver.GMLevel);
-		bitStream.Write(eChatMessageResponseCode::ReceivedNewWhisper);
-		bitStream.Write(message);
+	BitStreamUtils::WriteHeader(bitStream, eConnectionType::CHAT, eChatMessageType::PRIVATE_CHAT_MESSAGE);
+	bitStream.Write(sender.playerID);
+	bitStream.Write(eChatChannel::TEAM);
```
Comment (EmosewaMC) [inline dChatServer/ChatPacketHandler.cpp:450]: This function is called SendPrivateChatmessage, but the chat channel is Team.  Why was this changed from what it was previously?

File: dChatServer/PlayerContainer.h [hunk 12]
```diff
@@ -42,7 +43,8 @@ struct PlayerData {
 	std::string playerName;
 	std::vector<FriendData> friends;
 	std::vector<IgnoreData> ignoredPlayers;
-	uint8_t GMLevel = 0;
+	// Unused TODO: make use of this
+	eGameMasterLevel GMLevel = eGameMasterLevel::CIVILIAN;
```
Comment (EmosewaMC) [inline dChatServer/PlayerContainer.h:47]: I would remove the include, forward declare and move the initializer to be defined in PlayerContainer.cpp to reduce the include clutter

File: dChatServer/ChatPacketHandler.cpp [hunk 13]
```diff
@@ -398,72 +386,83 @@ void ChatPacketHandler::HandlePrivateChatMessage(Packet* packet) {
...
-		bitStream.Write<uint16_t>(0); // sourceID
-		bitStream.Write(sender.GMLevel);
-		bitStream.Write(LUWString(receiver.playerName));
-		bitStream.Write(receiver.GMLevel);
-		bitStream.Write(eChatMessageResponseCode::Sent);
-		bitStream.Write(message);
+	if (receiver.isFTP) {
+		SendPrivateChatMessage(sender, receiver, sender, message, eChatMessageResponseCode::RECEIVERFREETRIAL);
+		return;
+	}
```
Comment (EmosewaMC) [inline dChatServer/ChatPacketHandler.cpp:427]: Same thing here.  I dont like the prospect of actually starting to add code to support FTP features and locks.

File: dChatServer/ChatPacketHandler.cpp (hunk 11 above)
Comment (aronwk-aaron) [reply dChatServer/ChatPacketHandler.cpp:450]: yep, it should take in the channel to send on as well. oversight on my part
teams is technically a type of private chat message as well

File: dChatServer/ChatPacketHandler.cpp (hunk 10 above)
Comment (aronwk-aaron) [reply dChatServer/ChatPacketHandler.cpp:422]: fair, got carried away covering every case

**Review by EmosewaMC — APPROVED** [review]

**Review by EmosewaMC — APPROVED** [review]

**Review by Xiphoseer — APPROVED** [review]
